### PR TITLE
feat: change approx percentile/median UDFs to return floats

### LINKED ISSUE #18092: Refactor away usage of `NUMERICS`/`INTEGERS` in `datafusion/expr-common/src/type_coercion/aggregates.rs`
https://github.com/apache/datafusion/blob/264030cca76d0bdb4d8809f252b422e72624a345/datafusion/expr-common/src/type_coercion/aggregates.rs#L44-L66

Refactor these away in favour of using the TypeSignature API: https://github.com/apache/datafusion/blob/264030cca76d0bdb4d8809f252b422e72624a345/datafusion/expr-common/src/signature.rs

Need to take care for functions that will be affected, especially as for `NUMERICS` it doesn't contain `Float16` or `Decimal`s which `TypeSignature::Numeric` does include. Consider what changes might be necessary for downstream functions.

### PR DESCRIPTION
**Pull request by theirix**
## Which issue does this PR close?

- Closes #18092.

## Rationale for this change

1. Migrating to the modern TypeSignature API: [264030c/datafusion/expr-common/src/signature.rs](https://github.com/apache/datafusion/blob/264030cca76d0bdb4d8809f252b422e72624a345/datafusion/expr-common/src/signature.rs)

2. Coercing types of `approx_percentile_cont`, `approx_percentile_cont_with_weight`, `approx_median` to floats. It matches PostgreSQL, DuckDB, and ClickHouse behaviour, except for Spark.

## What changes are included in this PR?

- Port remaining UDFs (approx_percentile_cont, approx_percentile_cont_with_weight, approx_median, stub functions) to signature APIs
- Deprecate INTEGERS and NUMERICS arrays in favour of using the TypeSignature API
- They are not removed yet, but marked as deprecated to avoid breaking downstream
- Fix up a SLT for approx_percentile_cont, approx_median to make sure it returns a float

## Are these changes tested?

- Tests are passing
- Updated tests to expect floats in return types

## Are there any user-facing changes?

- Signatures of `approx_percentile_cont`, `approx_percentile_cont_with_weight`, `approx_median` changed, so they now return floats instead of integers (as seen in tests)

### CODE REVIEW

File: datafusion/expr-common/src/type_coercion/aggregates.rs [hunk 1]
```diff
@@ -22,6 +22,7 @@ use datafusion_common::{Result, internal_err, plan_err};
 
 // TODO: remove usage of these (INTEGERS and NUMERICS) in favour of signatures
 //       see https://github.com/apache/datafusion/issues/18092
+#[deprecated(since = "54.0.0", note = "Use functions signatures")]
```
Comment (Jefffrey) [inline datafusion/expr-common/src/type_coercion/aggregates.rs:23]: Do we need to remove the comment above too?

File: datafusion/sqllogictest/test_files/aggregate.slt [hunk 2]
```diff
@@ -144,28 +144,22 @@ statement error DataFusion error: Schema error: Schema contains duplicate unqual
...
 
 # csv_query_approx_percentile_cont_with_histogram_bins
 statement error DataFusion error: Error during planning: Tdigest max_size value for 'APPROX_PERCENTILE_CONT' must be UInt > 0 literal \(got data type Int64\)\.
 SELECT c1, approx_percentile_cont(0.95, -1000) WITHIN GROUP (ORDER BY c3) AS c3_p95 FROM aggregate_test_100 GROUP BY 1 ORDER BY 1
 
-statement error Failed to coerce arguments to satisfy a call to 'approx_percentile_cont' function
+statement error Function 'approx_percentile_cont' failed to match any signature
 SELECT approx_percentile_cont(0.95, c1) WITHIN GROUP (ORDER BY c3) FROM aggregate_test_100
 
-statement error DataFusion error: Error during planning: Failed to coerce arguments to satisfy a call to 'approx_percentile_cont' function: coercion from Int16, Float64, Float64 to the signature OneOf(.*) failed(.|\n)*
```
Comment (Jefffrey) [inline datafusion/sqllogictest/test_files/aggregate.slt:193]: Should we be removing these tests?

File: datafusion/expr-common/src/signature.rs [hunk 3]
```diff
@@ -595,6 +594,20 @@ impl Display for ArrayFunctionArgument {
     }
 }
 
+static NUMERICS: &[DataType] = &[
```
Comment (Jefffrey) [inline datafusion/expr-common/src/signature.rs:598]: What do we still need this for? Especially considering it seems to be missing decimals

**Review by Jefffrey — COMMENTED** [review]
Thanks for taking this up. Some of my concerns which held me off on this effort was the impact on error messaging (though I recall there was a recent PR to clean this up #20605) and if this deals with null type inputs correctly now (see #19458)

Also it seems some of the return types are changing now, I think we should call this out?

File: datafusion/sqllogictest/test_files/aggregate.slt (hunk 2 above)
Comment (theirix) [reply datafusion/sqllogictest/test_files/aggregate.slt:193]: They actually pass now because the second parameter is coercible.

File: datafusion/expr-common/src/signature.rs (hunk 3 above)
Comment (theirix) [reply datafusion/expr-common/src/signature.rs:598]: It is used only for example - TypeSignature::get_example_types. I removed it and cherry-picked three of them for illustrative purposes.

UPDATE: it isn't just example types. I'd suggest refactoring it as a part of https://github.com/apache/datafusion/issues/14761

File: datafusion/expr-common/src/type_coercion/aggregates.rs (hunk 1 above)
Comment (theirix) [reply datafusion/expr-common/src/type_coercion/aggregates.rs:23]: Done. Deprecation is enough.

**Comment by theirix** [comment]
> Thanks for taking this up. Some of my concerns which held me off on this effort was the impact on error messaging (though I recall there was a recent PR to clean this up #20605) and if this deals with null type inputs correctly now (see #19458)
> 
> Also it seems some of the return types are changing now, I think we should call this out?

Thank you for the review!

Added more tests for null - seems like it's consistent. The return value of these UDFs was float before (it is seen in UDF doc and also in other SQL engines), so it's reasonable to expect floats in tests too. Do you mean this change of types?

File: datafusion/functions-aggregate/src/approx_percentile_cont.rs [hunk 4]
```diff
@@ -132,22 +132,45 @@ impl Default for ApproxPercentileCont {
...
-                    DataType::Float64,
-                    int.clone(),
-                ]))
-            }
-        }
-        Self {
-            signature: Signature::one_of(variants, Volatility::Immutable),
-        }
+        // Additionally accept an integer number of centroids for T-Digest
+        let signature = Signature::one_of(
```
Comment (Jefffrey) [inline datafusion/functions-aggregate/src/approx_percentile_cont.rs:133]: Since we're now coercing to floats, we can remove some of the implementation code that handles integer types

**Review by Jefffrey — COMMENTED** [review]
> Added more tests for null - seems like it's consistent. The return value of these UDFs was float before (it is seen in UDF doc and also in other SQL engines), so it's reasonable to expect floats in tests too. Do you mean this change of types?

Yes, just referring to how it seems we used to return the input type but now we'll always coerce to float. We can see the return type changing in the SLTs, so would be good to highlight this in the title + PR body (since now we're going beyond a simple refactor)

File: datafusion/functions-aggregate/src/approx_percentile_cont.rs (hunk 4 above)
Comment (theirix) [reply datafusion/functions-aggregate/src/approx_percentile_cont.rs:133]: Agree, removed some code. It also affected `approx_median`, piggybacking on `approx_percentile_cont` - updated signature and tests as well. 

Clarified the scope of this PR in description - it changes more than expected

**Review by Jefffrey — APPROVED** [review]
Makes sense to change these to return floats since they are meant to be continuous instead of discrete 👍

**Comment by theirix** [comment]
Yes, it makes sense. Thank you for the review!

**Comment by Jefffrey** [comment]
With the output type changing, I wonder if that is a significant enough change to warrant an entry in the upgrade guide 🤔

**Comment by theirix** [comment]
> With the output type changing, I wonder if that is a significant enough change to warrant an entry in the upgrade guide 🤔

I agree. Added to the PR description and to the 54.0.0 upgrade docs. We can also mention it in #21080

**Comment by Jefffrey** [comment]
Thanks @theirix
